Header.c

```diff
@@ -22,135 +23,156 @@ in the source distribution for its full text.
+void Header_setLayout(Header* this, HeaderLayout hLayout) {
+   size_t oldColumns = HeaderLayout_getColumns(this->headerLayout);
+   size_t newColumns = HeaderLayout_getColumns(hLayout);
 
-      const Vector* vec = this->columns[col];
-      int len = Vector_size(vec);
+   this->headerLayout = hLayout;
 
-      colSettings->names = xCalloc(len + 1, sizeof(char*));
-      colSettings->modes = xCalloc(len, sizeof(int));
-      colSettings->len = len;
+   if (newColumns == oldColumns)
+      return;
 
-      for (int i = 0; i < len; i++) {
-         const Meter* meter = (Meter*) Vector_get(vec, i);
-         char* name;
-         if (meter->param && As_Meter(meter) == &DynamicMeter_class) {
-            const char* dynamic = DynamicMeter_lookup(this->pl->dynamicMeters, meter->param);
-            xAsprintf(&name, "%s(%s)", As_Meter(meter)->name, dynamic);
-         } else if (meter->param && As_Meter(meter) == &CPUMeter_class) {
-            xAsprintf(&name, "%s(%u)", As_Meter(meter)->name, meter->param);
-         } else {
-            xAsprintf(&name, "%s", As_Meter(meter)->name);
+   if (newColumns > oldColumns) {
+      this->columns = xReallocArray(this->columns, newColumns, sizeof(Vector*));
+      for (size_t i = oldColumns; i < newColumns; i++)
+         this->columns[i] = Vector_new(Class(Meter), true, DEFAULT_SIZE);
+   } else {
+      // move meters from to-be-deleted columns into last one
+      for (size_t i = newColumns; i < oldColumns; i++) {
+         for (int j = this->columns[i]->items - 1; j >= 0; j--) {
+            Vector_add(this->columns[newColumns - 1], Vector_take(this->columns[i], j));
          }
-         colSettings->names[i] = name;
-         colSettings->modes[i] = meter->mode;
+         Vector_delete(this->columns[i]);
       }
+      this->columns = xReallocArray(this->columns, newColumns, sizeof(Vector*));
    }
+
+   Header_calculateHeight(this);
 }
 
-bool Header_addMeterByName(Header* this, const char* name, int column) {
+static void Header_addMeterByName(Header* this, const char* name, MeterModeId mode, unsigned int column) {
+   assert(column < HeaderLayout_getColumns(this->headerLayout));
+
    Vector* meters = this->columns[column];
 
-   char* paren = strchr(name, '(');
+   const char* paren = strchr(name, '(');
    unsigned int param = 0;
+   size_t nameLen;
    if (paren) {
-      char* end, dynamic[32] = {0};
       int ok = sscanf(paren, "(%10u)", &param); // CPUMeter
       if (!ok) {
+         char* end, dynamic[32] = {0};
          if (sscanf(paren, "(%30s)", dynamic)) { // DynamicMeter
             if ((end = strrchr(dynamic, ')')) == NULL)
-               return false;    // indicate htoprc parse failure
+               return;    // indicate htoprc parse failure
             *end = '\0';
             if (!DynamicMeter_search(this->pl->dynamicMeters, dynamic, &param))
-               return false;    // indicates name lookup failure
+               return;    // indicates name lookup failure
```

Comment:
This comment doesn't make sense anymore  (no return value but its indicating failure?)  Similarly on line 90, three lines prior.

Reply:
I saw those too yesterday. Will fix them later, unless someone else changes them first … ;-)

Reply:
I think the comments still make sense to explain the error conditions that trigger the (silent) return.

Reply:
What about dropping the `indicates`?

Reply:
yeah, still serves the purpose